### DISCUSSION TITLE
Issue with missing content and blank pages

### DISCUSSION THREAD

**Discussion by mrdo23** [hardkoded/puppeteer-sharp#2409] [Q&A]
Hi,
I am running some code to take an HTML string, fetch some external CSS files and then save the formatted output as a PDF.
Unfortunately, I regularly receive incomplete output or blank pages, despite the same raw HTML being passed to puppeteer-sharp.
As per various SO posts, I have added 'WaitUntilNavigation.Networkidle0' and 'WaitForSelectorAsync' code to see if that helps but the problem persists.
Does anyone have any suggestions that may help? Thanks in advance.
```
using (Browser browser = await Puppeteer.LaunchAsync(launchOptions))
            {
                using (var page = await browser.NewPageAsync())
                {

                    await page.SetContentAsync(pageHTML, new NavigationOptions { WaitUntil = new[] { WaitUntilNavigation.Networkidle0 } });

                    // pull in main css from site for formatting
                    AddTagOptions css = new AddTagOptions() { Url = "https://MYURL.COM/main.css" };
                    await page.AddStyleTagAsync(css);
                    // pull in font-awesome css from site for formatting - linking to our own copy doesn't seem to work
                    AddTagOptions cssFA = new AddTagOptions() { Url = "https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.14.0/css/all.min.css" };
                    await page.AddStyleTagAsync(cssFA);


                    await page.AddStyleTagAsync(new AddTagOptions { Content = inlineCSS });

                    await page.GetContentAsync();

                    await page.WaitForSelectorAsync("#lastelement");

                    var file = await page.PdfDataAsync(pdfOptions);

                    var fileName = "result.pdf";

                    return new FileContentResult(file, "application/pdf")
                    {
                        FileDownloadName = fileName
                    };
                }
            }
```

**Comment by kblok** [comment]
As you can see. There is no silver bullet for this.
One thing you can to is adding a wait for expression and wait for some CSS style to be applied on an element.

Something like this :

```
await Page.WaitForExpressionAsync("window.getComputedStyle(document.querySelector('#someSelector')).color === rgb(132, 141, 151)");
```

Or wait for an element to get some height. I bet the solution is on some wait for expression that can do the trick for you.

**Comment by mrdo23** [comment]
Damn, was hoping that you had a cupboard full of silver bullets!

Let me play around with your suggestions and see if I can see any improvement.

Thanks for getting back to me - much appreciated.

**Comment by mrdo23** [comment]
I have resorted to using a delay of 3s after calling page.SetContentAsync - are you aware of any alternative to this?

**Reply by kblok** [reply]
I'm not a fan of fixed delays. What `WaitForExpressionAsync` have you tried?

**Reply by satviktechie1986** [reply]
not worked

**Comment by mrdo23** [comment]
Neither am I - slightly embarassed to have it in my code.

The issue is that I'm finding is that it's the content (SetContentAsync ) that is missing - my current WaitForExpressionAsync calls were passing as they reply on the css and check the formatting accordingly. I am injecting some text into the content - is there an expression to check for specific text in the content?
